File: NumDisappeared.java
/*448. Find All Numbers Disappeared in an Array
Time Complexity: O(n)
Space Complexity: O(n)
*/

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        HashSet<Integer> set = new HashSet<>();
        List<Integer> res = new ArrayList<>();

        for(int i=0; i<nums.length; i++)
            set.add(nums[i]);

        //Iterated from 1 to N since the length is max 1 to N [1,N] inclusive
        for(int i=1; i<=nums.length; i++){
            if(!set.contains(i))
                res.add(i);
        }

        return res;
    }
}


/*448. Find All Numbers Disappeared in an Array
Time Complexity: O(n)
Space Complexity: O(1)
*/

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> result = new ArrayList<>();

        for(int i=0; i<nums.length; i++){
            int index = Math.abs(nums[i]) - 1;

            if(nums[index] > 0)
                nums[index] *= -1;
        }

        for(int i=0; i<nums.length; i++){
            if(nums[i] > 0)
                result.add(i+1);
        }
        return result;
    }
}

File: num_disappeared.py
# Time Complexity : O(N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach
# Easy approach just make the indexes which are present to negative
# so the positive ones are missing. 


class Solution(object):
    def findDisappearedNumbers(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = []
        n = len(nums)
        for i in range(0,n):
            index = abs(nums[i])-1
            if(nums[index] > 0):
                nums[index]*=-1
        for i in range(0,n):
            if(nums[i]>0):
                res.append(i+1)
        return res
        
